typescript/xstate/src/lib.ts

```diff
@@ -0,0 +1,360 @@
+    _parent: actorRef._parent === undefined ? undefined : serialiseActorRef(actorRef._parent)
+  }
+}
+
+type SerialisableScheduledEvent = {
+  id: string;
+  event: EventObject;
+  startedAt: number;
+  delay: number;
+  source: SerialisableActorRef;
+  target: SerialisableActorRef;
+  uuid: string;
+}
+
+async function createSystem<T extends ActorSystemInfo>(ctx: restate.RpcContext, api: XStateApi<ActorLogicFrom<T>>, systemName: string): Promise<RestateActorSystem<T>> {
+  const events = await ctx.get<{ [key: string]: SerialisableScheduledEvent }>("events") ?? {}
+  const childrenByID = await ctx.get<{ [key: string]: SerialisableActorRef }>("children") ?? {}
+
+  let idCounter = 0;
+  const children = new Map<string, AnyActorRef>();
+  const keyedActors = new Map<keyof T['actors'], AnyActorRef | undefined>();
+  const reverseKeyedActors = new WeakMap<AnyActorRef, keyof T['actors']>();
+  const observers = new Set<Observer<InspectionEvent>>();
+
+  const scheduler = {
+    schedule(_source: AnyActorRef, _target: AnyActorRef, event: EventObject, delay: number, id: string | undefined): void {
+      if (id === undefined) {
+        id = ctx.rand.random().toString(36).slice(2)
+      }
+
+      const {source, target} = {source: serialiseActorRef(_source), target: serialiseActorRef(_target)}
+
+      console.log("schedule from", source.id, "to", target.id, "with id", id, "and delay", delay)
+
+      const scheduledEvent: SerialisableScheduledEvent = {
+        source,
+        target,
+        event,
+        delay,
+        id,
+        startedAt: Date.now(),
+        uuid: ctx.rand.uuidv4(),
+      };
+      const scheduledEventId = createScheduledEventId(source, id);
+      if (scheduledEventId in events) {
+        console.log("Ignoring duplicated schedule from", source.id, "to", target.id)
+        return
+      }
+
+      events[scheduledEventId] = scheduledEvent;
+
+      ctx.sendDelayed(api.actor, delay).send(systemName, {scheduledEvent, source, target, event})
+      ctx.set("events", events)
+    },
+    cancel(source: AnyActorRef, id: string): void {
+      console.log("cancel schedule from", source.id, "with id", id)
+
+      const scheduledEventId = createScheduledEventId(source, id);
+
+      delete events[scheduledEventId];
+      ctx.set("events", events)
+    },
+    cancelAll(actorRef: AnyActorRef): void {
+      console.log("cancel all for", actorRef.id)
+
+      for (const scheduledEventId in events) {
+        const scheduledEvent = events[scheduledEventId];
+        if (scheduledEvent.source.sessionId === actorRef.sessionId) {
+          delete events[scheduledEventId];
+        }
+      }
+      ctx.set("events", events)
+    },
+  }
+
+  const system: RestateActorSystem<T> = {
+    ctx,
+    api,
+    systemName,
+
+    _bookId: () => ctx.rand.uuidv4(),
+    _register: (sessionId, actorRef) => {
+      if (actorRef.id in childrenByID) {
+        // rehydration case; ensure session ID maintains continuity
```

Comment:
general question; is my use of `sessionid` right? i treat it as a particular execution of an actor, so we want the id to be unique, but also deterministic on rehydrate (ie, on the next event where we start up the state machine all over again). thats why I deliberately store session ids in state, and then correlate them back with the rehydrated children using `id` in order to set the same session ids